### PR TITLE
Add NNX transforms `nnx.while_loop` and `nnx.switch`

### PR DESCRIPTION
**Pull request by IvyZX**
* Implemented `nnx.switch` similar to `nnx.cond`

* Implemented `nnx.while_loop`

  * Due to constraints of `jax.lax.while_loop`, no reference structure change is allowed inside `nnx.while_loop` for NNX objects.

### CODE REVIEW

File: flax/nnx/transforms/transforms.py [hunk 1]
```diff
@@ -160,3 +164,141 @@ def cond(
...
+               body_fun: tp.Callable[[T], T],
+               init_val: T) -> T:
+  """NNX transform of `jax.lax.while_loop`.
+
+  See: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html
+
+  Caution: for the NNX internal reference tracing mechanism to work, you cannot
+  change the reference structure of `init_val` inside `body_fun`.
+
+  Args:
```
Comment (8bitmp3) [inline flax/nnx/transforms/transforms.py:282]: Maybe:
```diff
  Args:
    cond_fun: a function for the continue condition of the while loop, taking a
-       single input of type `T` and outputs a boolean.
+       single input of type `T` and outputting a boolean.
    body_fun: a function that takes an input of type `T` and outputs an `T`.
      Note that both data and modules of `T` must have the same reference
      structure between inputs and outputs.
    init_val: the initial input for cond_fun and body_fun. Must be of type `T`.
```

Comment (IvyZX) [reply flax/nnx/transforms/transforms.py:282]: Thanks! Fixed.

File: flax/nnx/transforms/transforms.py [hunk 2]
```diff
@@ -160,3 +164,141 @@ def cond(
...
+    out = self.f(val)
+    return out
+
+
+def _add_fake_index_mapping(tree: tp.Any):
+  def per_node_state(ns: extract.NodeStates | tp.Any):
+    global_index_mapping = {}
+    if not isinstance(ns, extract.NodeStates):
+      return ns
+    assert isinstance(ns._graphdef, graph.NodeDef)
```
Comment (cgarciae) [inline flax/nnx/transforms/transforms.py:205]: When you have a repeated NNX object in the pytree processed by `to_tree` you will get a `NodeRef` here. You can just return in that case. Fix:

```suggestion
    if not isinstance(ns._graphdef, graph.NodeDef):
      return ns
```
I tested this example that fails in this branch.
```python
m = nnx.Linear(10, 10, rngs=nnx.Rngs(0))

def body_fn(val):
  count, m, _ = val
  return count + 1, m, m

count, m, _ = nnx.while_loop(
  lambda val: val[0] < 2,
  body_fn,
  (0, m, m),
)
```

File: flax/nnx/transforms/transforms.py [hunk 3]
```diff
@@ -160,3 +164,141 @@ def cond(
...
+  return jax.tree.map(per_node_state, tree,
+                      is_leaf=lambda x: isinstance(x, extract.NodeStates))
+
+
+def _remove_index_mapping(tree: tp.Any):
+  '''Remove a fake index_mapping for the input to match that of the output.'''
+  def per_node_state(ns: extract.NodeStates | tp.Any):
+    if not isinstance(ns, extract.NodeStates):
+      return ns
+    assert isinstance(ns._graphdef, graph.NodeDef)
```
Comment (cgarciae) [inline flax/nnx/transforms/transforms.py:232]: Same as before.

File: flax/nnx/transforms/transforms.py [hunk 4]
```diff
@@ -160,3 +164,141 @@ def cond(
...
+    return pure_out
+
+
+@graph.update_context('while_loop')
+def while_loop(cond_fun: tp.Callable[[T], tp.Any],
+               body_fun: tp.Callable[[T], T],
+               init_val: T) -> T:
+  """NNX transform of `jax.lax.while_loop`.
+
+  See: https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html
```
Comment (cgarciae) [inline flax/nnx/transforms/transforms.py:277]: More friendly link.
```suggestion
  See `jax.lax.while_loop <https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.while_loop.html>`_.
```

Comment (cgarciae) [reply flax/nnx/transforms/transforms.py:277]: Or maybe we link be above mention?

File: flax/nnx/transforms/transforms.py [hunk 5]
```diff
@@ -160,3 +164,141 @@ def cond(
...
+
```
Comment (cgarciae) [inline flax/nnx/transforms/transforms.py:278]: Can we add a simple example usage here?

**Review by cgarciae — APPROVED** [review]

**Comment by cgarciae** [comment]
BTW: Can we move `while_loop` to `iteration.py`? Been meaning to remove `transforms.py` which previously contained all the transforms in single file that became too big.

File: flax/nnx/transforms/transforms.py (hunk 5 above)
Comment (IvyZX) [reply flax/nnx/transforms/transforms.py:278]: Done.

File: flax/nnx/transforms/transforms.py (hunk 1 above)
Comment (8bitmp3) [reply flax/nnx/transforms/transforms.py]: and thank you!

**Pull request merged**